state-sync/src/shared_components.rs

```diff
@@ -0,0 +1,241 @@
+
+use crate::logging::{LogEntry, LogSchema};
+use anyhow::{format_err, Result};
+use diem_logger::prelude::*;
+use diem_types::{
+    epoch_change::Verifier, epoch_state::EpochState, ledger_info::LedgerInfoWithSignatures,
+    transaction::Version,
+};
+use executor_types::ExecutedTrees;
+use std::{collections::BTreeMap, ops::Bound::Included};
+
+// DS to help sync requester to keep track of ledger infos in the future
+// if it is lagging far behind the upstream node
+// Should only be modified upon local storage sync
+pub(crate) struct PendingLedgerInfos {
+    // In-memory store of ledger infos that are pending commits
+    // (k, v) - (LI version, LI)
+    pending_li_queue: BTreeMap<Version, LedgerInfoWithSignatures>,
+    // max size limit on `pending_li_queue`, to prevent OOM
+    max_pending_li_limit: usize,
+    // target li
+    target_li: Option<LedgerInfoWithSignatures>,
+}
+
+impl PendingLedgerInfos {
+    pub(crate) fn new(max_pending_li_limit: usize) -> Self {
+        Self {
+            pending_li_queue: BTreeMap::new(),
+            max_pending_li_limit,
+            target_li: None,
+        }
+    }
+
+    /// Adds `new_li` to the queue of pending LI's
+    pub(crate) fn add_li(&mut self, new_li: LedgerInfoWithSignatures) {
+        if self.pending_li_queue.len() >= self.max_pending_li_limit {
+            warn!(
+                LogSchema::new(LogEntry::ProcessChunkResponse),
+                "pending LI store reached max capacity {}, failed to add LI {}",
+                self.max_pending_li_limit,
+                new_li
+            );
+            return;
+        }
+
+        // update pending_ledgers if new LI is ahead of target LI (in terms of version)
+        let target_version = self
+            .target_li
+            .as_ref()
+            .map_or(0, |li| li.ledger_info().version());
+        if new_li.ledger_info().version() > target_version {
+            self.pending_li_queue
+                .insert(new_li.ledger_info().version(), new_li);
+        }
+    }
+
+    pub(crate) fn update(&mut self, sync_state: &SyncState, chunk_limit: u64) -> Result<()> {
+        let highest_committed_li = sync_state.committed_version();
+        let highest_synced = sync_state.synced_version();
+
+        // prune any pending LIs that are older than the latest local synced version
+        let prune_version = highest_synced
+            .checked_add(1)
+            .ok_or_else(|| format_err!("Prune version has overflown!"))?;
+        self.pending_li_queue = self.pending_li_queue.split_off(&prune_version);
+
+        // pick target LI to use for sending ProgressiveTargetType requests.
+        self.target_li = if highest_committed_li == highest_synced {
+            // try to find LI with max version that will fit in a single chunk
+            let highest_version = highest_synced
+                .checked_add(chunk_limit)
+                .ok_or_else(|| format_err!("Highest version has overflown!"))?;
```

Comment:
I'm not sure I follow you.. the error is only if we overflow on the adds (in which case, it makes sense to halt because we've hit some fundamental limits and it's not clear what should happen next).